File: content-solutions/graph-advanced/uf-group-sizes.java
import java.util.*;

class Solution {
    int[] parent, size;

    void init(int n) {
        parent = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    int find(int x) {
        int root = x;
        while (parent[root] != root) root = parent[root];
        while (x != root) {
            int next = parent[x];
            parent[x] = root;
            x = next;
        }
        return root;
    }

    boolean union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;
        if (size[ra] < size[rb]) {
            int t = ra;
            ra = rb;
            rb = t;
        }
        parent[rb] = ra;
        size[ra] += size[rb];
        return true;
    }

    public int[] solution(int n, int[][] pairs) {
        init(n);
        for (int[] p : pairs) union(p[0], p[1]);
        List<Integer> sizes = new ArrayList<>();
        for (int i = 0; i < n; i++) if (find(i) == i) sizes.add(size[i]);
        sizes.sort(Collections.reverseOrder());
        int[] answer = new int[sizes.size()];
        for (int i = 0; i < answer.length; i++) answer[i] = sizes.get(i);
        return answer;
    }
}

File: content-solutions/graph-advanced/uf-group-sizes.js
class DSU {
  constructor(n) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.size = new Array(n).fill(1);
  }
  find(x) {
    let root = x;
    while (this.parent[root] !== root) root = this.parent[root];
    while (x !== root) {
      const next = this.parent[x];
      this.parent[x] = root;
      x = next;
    }
    return root;
  }
  union(a, b) {
    let ra = this.find(a);
    let rb = this.find(b);
    if (ra === rb) return false;
    if (this.size[ra] < this.size[rb]) [ra, rb] = [rb, ra];
    this.parent[rb] = ra;
    this.size[ra] += this.size[rb];
    return true;
  }
}

function solution(n, pairs) {
  const dsu = new DSU(n);
  for (const [a, b] of pairs) dsu.union(a, b);
  const sizes = [];
  for (let i = 0; i < n; i++) if (dsu.find(i) === i) sizes.push(dsu.size[i]);
  return sizes.sort((a, b) => b - a);
}
